### PR TITLE
ch4: misc fixes for rma am/ofi with complex datatypes (fixing bugs found in armci-mpi)

### PR DESCRIPTION
**Pull request by minsii**

### CODE REVIEW

**Comment by minsii** [comment]
test:jenkins/ch4/ofi

**Comment by minsii** [comment]
test:jenkins/ch4/ofi

**Comment by minsii** [comment]
@gcongiu Looks like current dtpool does not cover a case where recv datatype (or RMA target datatype) is `contiguous` but `true_lb > 0`. See commit `ch4/ofi: add target_lb to remote addr at contig put` in this PR, the bug was never reported in per-review and nightly tests.

As an example of `contig + true_lb > 0` datatype, if I create an `indexed_block` type with `(count=1, blocklength=16, disps={4})`, it should be able to generate such a case. Could you please confirm ?

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch4/armci-mpi

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch4/armci-mpi

**Comment by gcongiu** [comment]
@minsii there is no starting offset implemented for datatypes in dtpools right now. All datatypes created in a pool start from offset `0`.

**Comment by pavanbalaji** [comment]
I think Min is asking if such a type can be created.  It’ll be useful for my PR testing too.

**Comment by gcongiu** [comment]
@pavanbalaji yes I can add a new layout to https://github.com/pmodels/mpich/pull/3678 to account for this case

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch4/armci-mpi

File: src/mpid/ch4/src/ch4r_rma_target_callbacks.c [hunk 1]
```diff
@@ -35,7 +35,7 @@ static int get_acc_iov_target_cmpl_cb(MPIR_Request * rreq);
 static int cswap_target_cmpl_cb(MPIR_Request * rreq);
 static int acc_target_cmpl_cb(MPIR_Request * rreq);
 static int get_acc_target_cmpl_cb(MPIR_Request * rreq);
-static int get_ack_target_cmpl_cb(MPIR_Request * rreq);
+static int get_ack_target_cmpl_cb(MPIR_Request * greq);
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_rma_target_callbacks.c]: naming here could be confusing given the existence of generalized requests. would suggest just `req` or `get_req`.

File: src/mpid/ch4/src/ch4r_rma_target_callbacks.c [hunk 2]
```diff
@@ -779,10 +779,14 @@ static void handle_acc_data(void **data, size_t * p_data_sz, int *is_contig, MPI
...
     MPIDI_Datatype_check_size(MPIDIG_REQUEST(rreq, req->areq.origin_datatype),
                               MPIDIG_REQUEST(rreq, req->areq.origin_count), data_sz);
+
+    /* The origin_data can be NULL only with no-op.
+     * TODO: when no-op is set, we do not need send origin_data at all. */
     if (data_sz) {
         p_data = MPL_malloc(data_sz, MPL_MEM_RMA);
         MPIR_Assert(p_data);
-    }
+    } else
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4r_rma_target_callbacks.c]: style nitpick - if one part of the if/else if/else uses {}, all parts should use them.

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch4/armci-mpi

**Comment by minsii** [comment]
@raffenet I have addressed all comments. It is ready for your check.

**Pull request merged**